### PR TITLE
Add unwrap_or_raise()

### PR DESCRIPTION
**Pull request by ileixe**
Fix: #93

### CODE REVIEW

**Comment by wbolster** [comment]
the basics here look good 👍🏼  but the API feels not so nice. i think we can do better, but let's keep that discussion in #93

**Comment by wbolster** [comment]
💡 just brainstorming…

i don't quite like how this proposed api makes it almost inevitable to use a lambda construct for something that's supposed to do something simple. (fwiw, i don't like lambdas in general, in python at least.)

maybe this could handle `Type[BaseException]` in a special way? so that it can be either a callable (such as a lambda) or an exception class directly?

``` python
res.unwrap_or_raise(ValueError)  # raise ValueError()
```

however, nice errors, which this is about in the first place, typically require more than just an exception class, e.g. a message:


``` python
res.unwrap_or_raise(ValueError, "foo")  # raise ValueError("foo")
```

but sometimes it's not just a string, so maybe also this?

``` python
res.unwrap_or_raise(MyException, *args, **kwargs)  # raise MyException(*args, **kwargs)
```

… but these may be hard to type correctly.

passing an exception _instance_ could work but would always construct it, even when it's not needed:

``` python
res.unwrap_or_raise(ValueError("foo"))  # hmmm :/
```

File: tests/test_result.py [hunk 1]
```diff
@@ -130,6 +130,14 @@ def test_unwrap_or_else() -> None:
     assert n.unwrap_or_else(str.upper) == 'NAY'
 
 
+def test_unwrap_or_raise() -> None:
+    o = Ok('yay')
+    n = Err('nay')
+    assert o.unwrap_or_raise(lambda e: Exception(e)) == 'yay'
+    with pytest.raises(Exception):
+        n.unwrap_or_raise(lambda e: Exception(e))
```
Comment (wbolster) [inline tests/test_result.py:138]: these can be plain `.unwrap_or_raise(Exception)`, in multiple places

while at it, please use `ValueError` or some other more specific exception, to make it clear that the passed in exception type is used for raising (and adjust the `pytest.raises()` call accordingly).

**Comment by wbolster** [comment]
this also needs docs in the readme

File: tests/test_result.py (hunk 1 above)
Comment (ileixe) [reply tests/test_result.py]: updated.

**Comment by ileixe** [comment]
> this also needs docs in the readme

Brief explanation for the new API added. I'm not sure it's enough to explain the API as it's a bit strange compared to others, so let me know if more detail needs.

File: README.rst [hunk 2]
```diff
@@ -235,6 +235,17 @@ A default value can be returned instead by using ``unwrap_or`` or ``unwrap_or_el
     >>> res2.unwrap_or_else(str.upper)
     'NAY'
 
+A exception can be raised from `Ok` or `Err`:
```
Comment (francium) [inline README.rst:238]: Just a small suggestion,
> The ``unwrap`` method will raised a ``UnwrapError``. A custom exception can be raised by using the ``unwrap_or_raise`` method in stead:

**Review by francium — COMMENTED** [review]
Looks good to me otherwise

File: tests/test_result.py [hunk 3]
```diff
@@ -130,6 +130,14 @@ def test_unwrap_or_else() -> None:
     assert n.unwrap_or_else(str.upper) == 'NAY'
 
 
+def test_unwrap_or_raise() -> None:
+    o = Ok('yay')
+    n = Err('nay')
+    assert o.unwrap_or_raise(ValueError) == 'yay'
+    with pytest.raises(ValueError):
```
Comment (wbolster) [inline tests/test_result.py:137]: can you add a test to verify that the exception instance actually received the error value?

💡 `pytest.raises` can give you the exception info, which has a `.value` attribute with the actual exception instance. something like this (untested):


``` python
    with pytest.raises(ValueError) as exc_info:
        ...

    exc = exc_info.value
    assert exc.args == ('nay,')
```

**Review by wbolster — CHANGES_REQUESTED** [review]
.

File: README.rst (hunk 2 above)
Comment (ileixe) [reply README.rst]: Thanks, applied.

File: tests/test_result.py (hunk 3 above)
Comment (ileixe) [reply tests/test_result.py]: Thanks, applied.

**Comment by francium** [comment]
Thanks @ileixe , let's see if it passes the tests and if so I'll test the changes locally and merge it if it all looks good.

**Review by francium — APPROVED** [review]

**Review by francium — CHANGES_REQUESTED** [review]
Forgot one final thing, can you please add a change log entry.
Here's an example if you need it, https://github.com/rustedpy/result/pull/82/files#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR16

**Comment by ileixe** [comment]
@francium Added. Thanks!

**Review by francium — APPROVED** [review]

**Comment by WesleyYue** [comment]
Does it make sense to add an overload for usage where an Exception is already the .err value?
```python
    def unwrap_or_raise(self) -> NoReturn:
        """
        The contained result is ``Err``, so raise the exception with the value.
        """
        raise self._value
```

Our code typically already uses a BaseException object for the err part of the tuple, so this would make more sense for us to adopt. This allows the inner functions to provide more detail about their exception in a pythonic way. For example, a FastAPI app might want to pass along `HTTPException(status_code=401, detail="User does not exist.")`, which would not be possible to use with unwrap_or_raise in its current state.

Happy to make a PR if there's interest. Though passing Exception in the err does go against the style that's currently in the readme.

**Comment by francium** [comment]
@WesleyYue can you please open a new issue for what you're describing. We can discuss it there rather than this old PR.

**Pull request merged**